service/worker/shadower/workflow.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2017-2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package shadower
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/worker"
+	"go.uber.org/cadence/workflow"
+
+	"github.com/uber/cadence/.gen/go/shadower"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+)
+
+const (
+	defaultScanWorkflowPageSize = 2000
+
+	// NOTE: do not simply change following values as it may result in workflow non-deterministic errors
+	defaultReplayConcurrency    = 1
+	defaultMaxReplayConcurrency = 50
+	defaultMaxShadowCountPerRun = 100000
+
+	defaultWaitDurationPerIteration = 5 * time.Minute
+)
+
+func register(worker worker.Worker) {
+	worker.RegisterWorkflowWithOptions(
+		shadowWorkflow,
+		workflow.RegisterOptions{Name: shadower.WorkflowName},
+	)
+	worker.RegisterActivity(verifyActiveDomainActivity)
+}
+
+func shadowWorkflow(
+	ctx workflow.Context,
+	params shadower.WorkflowParams,
+) (shadower.WorkflowResult, error) {
+	if err := validateAndFillWorkflowParams(&params); err != nil {
+		return shadower.WorkflowResult{}, err
+	}
+
+	lao := workflow.LocalActivityOptions{
+		ScheduleToCloseTimeout: time.Second * 5,
```

Comment:
what's the average expected execution time for the activity? I thought for a local one i should be up to few seconds.

Reply:
Yes this local activity only accesses domain cache and check if domain is active or not so it will be very fast. 

I don't have strong opinion on the timeout. 5s, 3s or 1s, I don't see a big difference here. Used 5 sec here just to be conservative, in case processing got slowed down when worker hosts are busy with other workflows, like archival or batch.